compiler/src/dotty/tools/dotc/transform/Staging.scala

```diff
@@ -77,373 +49,97 @@ class Staging extends MacroTransformWithImplicits {
-      def pickleAsTasty() = {
-        val meth =
-          if (isType) ref(defn.Unpickler_unpickleType).appliedToType(originalTp)
-          else ref(defn.Unpickler_unpickleExpr).appliedToType(originalTp.widen)
-        meth.appliedTo(
-          liftList(PickledQuotes.pickleQuote(body).map(x => Literal(Constant(x))), defn.StringType),
-          liftList(splices, defn.AnyType))
-      }
-      if (splices.nonEmpty) pickleAsTasty()
-      else if (isType) {
-        def tag(tagName: String) = ref(defn.QuotedTypeModule).select(tagName.toTermName)
-        if (body.symbol == defn.UnitClass) tag("UnitTag")
-        else if (body.symbol == defn.BooleanClass) tag("BooleanTag")
-        else if (body.symbol == defn.ByteClass) tag("ByteTag")
-        else if (body.symbol == defn.CharClass) tag("CharTag")
-        else if (body.symbol == defn.ShortClass) tag("ShortTag")
-        else if (body.symbol == defn.IntClass) tag("IntTag")
-        else if (body.symbol == defn.LongClass) tag("LongTag")
-        else if (body.symbol == defn.FloatClass) tag("FloatTag")
-        else if (body.symbol == defn.DoubleClass) tag("DoubleTag")
-        else pickleAsTasty()
-      }
-      else Staging.toValue(body) match {
-        case Some(value) => pickleAsValue(value)
-        case _ => pickleAsTasty()
-      }
-    }
-
-    /** If inside a quote, split the body of the splice into a core and a list of embedded quotes
-     *  and make a hole from these parts. Otherwise issue an error, unless we
-     *  are in the body of an inline method.
-     */
-    private def splice(splice: Select)(implicit ctx: Context): Tree = {
-      if (level > 1) {
-        val body1 = nested(isQuote = false).transform(splice.qualifier)
-        body1.select(splice.name)
-      }
-      else if (level == 1) {
-        val (body1, quotes) = nested(isQuote = false).split(splice.qualifier)
-        val tpe = outer.embedded.getHoleType(splice)
-        val hole = makeHole(body1, quotes, tpe).withSpan(splice.span)
-        // We do not place add the inline marker for trees that where lifted as they come from the same file as their
-        // enclosing quote. Any intemediate splice will add it's own Inlined node and cancel it before splicig the lifted tree.
-        // Note that lifted trees are not necessarily expressions and that Inlined nodes are expected to be expressions.
-        // For example we can have a lifted tree containing the LHS of an assignment (see tests/run-with-compiler/quote-var.scala).
-        if (splice.isType || outer.embedded.isLiftedSymbol(splice.qualifier.symbol)) hole
-        else Inlined(EmptyTree, Nil, hole).withSpan(splice.span)
+    protected def splice(splice: Select)(implicit ctx: Context): Tree = {
+      if (level >= 1) {
+        val body1 = transform(splice.qualifier)(spliceContext)
+        val splice1 = cpy.Select(splice)(body1, splice.name)
+        if (splice1.isType) splice1
+        else addSpliceCast(splice1)
```

Comment:
The following could be considered an unnecessary cast and optimized away,  right?

```scala
def test[T](x: Expr[T])(implicit t: Type[T]): Expr[T] = '{
    val x2: ~t = ~x
    x2
  }
```
->
```scala
val x2: ~(t) = ~(x).asInstanceOf[t.unary_~]
x2:t.unary_~
```

Reply:
Yes. Though I would do optimizations later as more refactoring will happen after this PR.